sign/dilithium/mode1/dilithium.go

```diff
@@ -0,0 +1,128 @@
+	common "github.com/cloudflare/circl/sign/dilithium/internal"
+	"github.com/cloudflare/circl/sign/dilithium/mode1/internal"
+)
+
+const (
+	// Size of seed for NewKeyFromSeed
+	SeedSize = common.SeedSize
+
+	// Size of a packed PublicKey
+	PublicKeySize = internal.PublicKeySize
+
+	// Size of a packed PrivateKey
+	PrivateKeySize = internal.PrivateKeySize
+
+	// Size of a signature
+	SignatureSize = internal.SignatureSize
+)
+
+// PublicKey is the type of Dilithium1 public key
+type PublicKey internal.PublicKey
+
+// PrivateKey is the type of Dilithium1 private key
+type PrivateKey internal.PrivateKey
+
+// GenerateKey generates a public/private key pair using entropy from rand.
+// If rand is nil, crypto/rand.Reader will be used.
+func GenerateKey(rand io.Reader) (*PublicKey, *PrivateKey, error) {
+	pk, sk, err := internal.GenerateKey(rand)
+	return (*PublicKey)(pk), (*PrivateKey)(sk), err
+}
+
+// NewKeyFromSeed derives a public/private key pair using the given seed.
+func NewKeyFromSeed(seed *[SeedSize]byte) (*PublicKey, *PrivateKey) {
+	pk, sk := internal.NewKeyFromSeed(seed)
+	return (*PublicKey)(pk), (*PrivateKey)(sk)
+}
+
+// SignTo signs the given message and writes the signature into signature.
+// It will panic if signature is not of length at least SignatureSize.
+func SignTo(sk *PrivateKey, msg []byte, signature []byte) {
+	internal.SignTo(
+		(*internal.PrivateKey)(sk),
+		msg,
+		signature,
+	)
+}
+
+// Sign signs the given message and returns the signature.
+func Sign(sk *PrivateKey, msg []byte) []byte {
+	var sig [SignatureSize]byte
+	SignTo(sk, msg, sig[:])
+	return sig[:]
+}
+
+// Verify checks whether the given signature by pk on msg is valid.
+func Verify(pk *PublicKey, msg []byte, signature []byte) bool {
+	return internal.Verify(
+		(*internal.PublicKey)(pk),
+		msg,
+		signature,
+	)
+}
+
+// Sets pk to the public key encoded in buf.
+func (pk *PublicKey) Unpack(buf *[PublicKeySize]byte) {
+	(*internal.PublicKey)(pk).Unpack(buf)
+}
+
+// Sets sk to the private key encoded in buf.
+func (sk *PrivateKey) Unpack(buf *[PrivateKeySize]byte) {
+	(*internal.PrivateKey)(sk).Unpack(buf)
+}
+
+// Packs the public key into buf.
+func (pk *PublicKey) Pack(buf *[PublicKeySize]byte) {
+	(*internal.PublicKey)(pk).Pack(buf)
+}
+
+// Packs the private key into buf.
+func (sk *PrivateKey) Pack(buf *[PrivateKeySize]byte) {
+	(*internal.PrivateKey)(sk).Pack(buf)
+}
+
+// Packs the public key.
+func (pk *PublicKey) MarshalBinary() ([]byte, error) {
+	var buf [PublicKeySize]byte
+	pk.Pack(&buf)
+	return buf[:], nil
+}
+
+// Packs the private key.
+func (sk *PrivateKey) MarshalBinary() ([]byte, error) {
+	var buf [PrivateKeySize]byte
+	sk.Pack(&buf)
+	return buf[:], nil
+}
+
+// Unpacks the public key from data.
+func (pk *PublicKey) UnmarshalBinary(data []byte) error {
+	if len(data) != PublicKeySize {
+		return errors.New("packed public key must be of PublicKeySize bytes")
+	}
+	var buf [PublicKeySize]byte
+	copy(buf[:], data)
+	pk.Unpack(&buf)
```

Comment:
No, there aren't.

Reply:
So is it true that any possible value in the array is mapped to a public key?

Reply:
I'm not sure, but I expect not.  But when you are given an RSA modulus `n` you also don't check whether it's actually the product of two primes — you simply compute with it because it doesn't matter.

Reply:
> My comment was in the direction of ensuring that the public/private key decoding can fail for some reasons. For example, here you are assuming that the encoding was performed using your function:
> 
> ```
> // Output coefficients of p are not normalized, but in [q-η,q+η] provided
> // buf was created using PackLeqEta.
> 
> // Beware, for arbitrary buf the coefficients of p might en up in
> // the interval [q-2^b,q+2^b] where b is the least b with η≤2^b.
> ```
> 
> but what happen if you receive the keys generated by other implementation, that it is expected to follow the same encoding process. Don't you think you should validate certain properties?

Earlier we were talking about the public key.  This is about `PackLeqEta` which is only used to unpack a private key.

The packed private key consists of `rho`, `key`, `s1`, `s2`, `t0` and `tr` of which the last two are redundant: they can be generated from the first values.  But that's as expensive as key generation itself.

Should we check that `t0` is generated correctly when unpacking the private key?  In that case we do not need the separate private key at all: we can just work with the seed that generated everything.

Checking whether `s1` and `s2` are below eta isn't very expensive, but it doesn't ensure at all that the private key is correctly generated or safe to use.  Perhaps it is prudent to document this fact and direct users to use `NewKeyFromSeed` instead.  (Which I'd rather call `DerivePrivateKeyFromSeed`.)

> > I'm not sure, but I expect not. But when you are given an RSA public key, then you also don't check whether the modulus n is the product of two primes — you simply compute with it because it doesn't matter.
> 
> [See this](https://github.com/google/boringssl/blob/master/crypto/fipsmodule/rsa/rsa.c#L712-L720)

That's a private key.  You asked about public keys.

Reply:
Continuing here #114